### PR TITLE
Use VS target version for package version of VS extensibility packages

### LINKED ISSUE #11394: [DCR]: NuGet.VisualStudio and NuGet.VisualStudio.Contracts to follow Visual Studio version numbers
### NuGet Product(s) Affected

Visual Studio Package Management UI, Visual Studio Package Manager Console

### Current Behavior

The packages `NuGet.VisualStudio` and `NuGet.VisualStudio.Contracts`, which define the interfaces for our Visual Studio extensibility, follow NuGet's version numbers. This follows Visual Studio's version numbers, but 11.0 versions behind. For example, NuGet 6.0 shipped in Visual Studio 2012 17.0. NuGet 5.11 shipped in Visual Studio 2019 16.11.

Additionally, the assemblies matched the build number, just like the rest of NuGet's assemblies. In other words, in build 6.0.0.280, all assemblies had the version 6.0.0.280. In NuGet build 5.11.0.10 has assembly version 5.11.0.10.

### Desired Behavior

The package version for the Visual Studio extensibility APIs should match the version (at least the major and minor version) of Visual Studio that those APIs are available in. For example, the package version for NuGet APIs that are available in dev 17.1 should be package version 17.1.x

Additionally, the assembly version for all builds in the same Major.Minor.Patch should use the same assembly version, to avoid binding redirect issues, similar to how Newtonsoft.Json use the same assembly version for all packages with the same major version.

### Additional Context

NuGet.VisualStudio 6.0.0 did not work when Visual Studio 2022 17.0 was first released, for reasons explained in this issue: https://github.com/NuGet/Home/issues/11367.

While the primary goal is to ensure this type of error can never happen again, we can take the opportunity to reduce confusion about which version of NuGet's APIs should be used in which version of Visual Studio.

### PR DESCRIPTION
**Pull request by zivkan**
Also make use same assembly version for all builds, so there are no
binding redirect issues when nupkgs published to nuget.org are a
different build number to what was inserted into VS.

## Bug

Fixes: https://github.com/NuGet/Home/issues/11394
Related: https://github.com/NuGet/Home/issues/11367

Regression? No

## Description

The reported problem was due to the fact that our assembly versions include the build number, which increments every build, and the packages we published to nuget.org were a higher build number than what was inserted into VS.  Our vsix adds binding redirects for our VS extensibility API packages/assemblies, but since the version we shipped in dev17.0.0 is lower than the assembly version we published to nuget.org, the binding redirect didn't apply.

So, the main goal of this PR is to ensure that the last digit of the assembly version is always zero, so it will no longer matter which specific builds of NuGet we insert into VS or publish to nuget.org. It'll all work out.

Our VS insertions require us to set the assembly version, so an extra parameter was added to `buildInfo.json`, so we can update the insertion pipeline to handle this correctly.

Finally, a secondary goal of this PR is to align the NuGet.VisualStudio and NuGet.VisualStudio.Contracts packages with the major and minor versions of Visual Studio that this build of NuGet is targeting, to make it easier for customers to know which version of the packages they should use, depending on which version of Visual Studio they are building their extension for. I'm going to go into more detail in a comment to this PR.

## PR Checklist

- [X] PR has a meaningful title
- [ ] PR has a linked issue.
- [X] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [X] N/A: No product changes, just assembly and package versions.

- **Documentation**
  
  - [X] Documentation PR or issue filled: https://github.com/NuGet/docs.microsoft.com-nuget/issues/2582
  - **OR**
  - [ ] N/A

### CODE REVIEW

**Comment by zivkan** [comment]
@NuGet/nuget-client This PR is making a change to the package version of `NuGet.VisualStudio` and `NuGet.VisualStudio.Contracts`. The goal is to have the package's major and minor version match the version of Visual Studio that the package is "designed for". However, there are some potential complications:

NuGet only closely follows Visual Studio's major and minor versions, not the patch versions. When a minor version goes GA, its patch number will be zero in both VS and NuGet: 17.1.0.  However, over time VS's patch number increases faster than NuGet's does (NuGet rarely services existing releases in the time I've been in the team, although looking at nuget.exe versions on nuget.org/downloads, there are some versions that were serviced multiple times).

The point is that it won't be uncommon for VS to be at a version number something like 17.1.4, while the version of NuGet.VisualStudio will be 17.1.0.  If we service NuGet, even if it's unrelated to VS, the package version will become 17.1.1, whereas VS will be 17.1.5.

In fact, we should never change APIs in a servicing fix, so theoretically we should only ever need to ship `x.y.0` versions of our VS API contracts, never `x.y.1` or higher. However, I don't think it's worthwhile to make changes to our build scripts and release pipelines to make that happen.

So, there's two downsides that I can see:

1. Small risk of confusion about why NuGet's VS extensibility packages have a different patch version to what VS does. However, I think there will be less confusion than today when the package's major version is 11 lower
2. These two packages no longer match the rest of our packages, or NuGet's "marketing" version.

What do you all think? Is this a good change?  I can revert the package version changes, keeping only the assembly version fix if that's preferred.

File: src/NuGet.Clients/NuGet.VisualStudio.Contracts/NuGet.VisualStudio.Contracts.csproj [hunk 1]
```diff
@@ -13,6 +13,9 @@
     <IncludeNuGetSharedFiles>true</IncludeNuGetSharedFiles>
     <RootNamespace>NuGet.VisualStudio.Contracts</RootNamespace>
     <Description>RPC contracts for NuGet's Visual Studio Service Broker extensibility APIs.</Description>
+    <!-- VS Extensibility APIs should use package version corresponding to VS version it targets, to reduce confusion. Also keep assembly version stable to minimize binding redirect issues. -->
+    <Version>$(NuGetSdkVsSemanticVersion)$(PreReleaseInformationVersion)</Version>
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.VisualStudio.Contracts/NuGet.VisualStudio.Contracts.csproj]: Should you use PackageVersion instead?

**Review by nkolev92 — COMMENTED** [review]
I'm supportive of this change. 
There's already a correlation between NuGet and VS version and due to the fact that these packages only matter in VS, might as well make the correlation even more explicit.

File: src/NuGet.Clients/NuGet.VisualStudio.Interop/NuGet.VisualStudio.Interop.csproj [hunk 2]
```diff
@@ -4,7 +4,6 @@
   <PropertyGroup>
     <TargetFramework>$(NETFXTargetFramework)</TargetFramework>
     <Shipping>true</Shipping>
-    <Pack>true</Pack>
```
Comment (dominoFire) [inline src/NuGet.Clients/NuGet.VisualStudio.Interop/NuGet.VisualStudio.Interop.csproj:7]: I think it looks unrelated to version. Can we do it in another PR specific to stop packing n.vs.interop ?

Comment (zivkan) [reply src/NuGet.Clients/NuGet.VisualStudio.Interop/NuGet.VisualStudio.Interop.csproj:7]: NuGet.VisualStudio.Interop isn't getting packed, never has. The property we use to determine packing is `PackProject`. This `Pack` property is unused in our build scripts.

I can remove it in another PR, but it's just clarifying the fact that the project isn't being packed, rather than stopping something that is currently packed from being packed.

File: src/NuGet.Clients/NuGet.VisualStudio.Contracts/NuGet.VisualStudio.Contracts.csproj [hunk 3]
```diff
@@ -13,6 +13,9 @@
     <IncludeNuGetSharedFiles>true</IncludeNuGetSharedFiles>
     <RootNamespace>NuGet.VisualStudio.Contracts</RootNamespace>
     <Description>RPC contracts for NuGet's Visual Studio Service Broker extensibility APIs.</Description>
+    <!-- VS Extensibility APIs should use package version corresponding to VS version it targets, to reduce confusion. Also keep assembly version stable to minimize binding redirect issues. -->
+    <PackageVersion>$(NuGetSdkVsSemanticVersion)$(PreReleaseInformationVersion)</PackageVersion>
+    <AssemblyVersion>$(NuGetSdkVsSemanticVersion).0</AssemblyVersion>
```
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.VisualStudio.Contracts/NuGet.VisualStudio.Contracts.csproj:18]: I am slowly catching up on this PR. If I understand this change correctly, going forward the assembly version of preview and GA packages will be the same. For e.g. the assembly version for the package versions `17.1.0-preview.1, 17.1.0-preview.2,17.1.0-preview.3 .... 17.1.0` will be `17.1.0.0`. Will there be any problems if we happen to introduce a breaking change during preview stages for the same VS target version given the fact that assembly version is going to the same?

Comment (zivkan) [reply src/NuGet.Clients/NuGet.VisualStudio.Contracts/NuGet.VisualStudio.Contracts.csproj:18]: > If I understand this change correctly, going forward the assembly version of preview and GA packages will be the same

Yes, that's the primary point of this PR, which prevents the problem reported by  NuGet/Home#11367. Changing package versions to match VS versions is a secondary goal, as described at the top of the PR comments.

> Will there be any problems if we happen to introduce a breaking change during preview stages for the same VS target version given the fact that assembly version is going to the same?

Consider what will happen if we keep incrementing the assembly version. We introduce a breaking change. The new build has a new assembly version, so our vsix's pkgdef creates a binding redirect to the newer assembly version. Other VS extensions that were compiled against lower assembly versions get redirected to the higher assembly version anyway.

So, there's no change in end result, it just protects us from assembly redirect version range problems.

Also remember that the packages that other VS extension authors use contain NuGet's interfaces only. If we introduce a bug in our implementation, using the old package version isn't going to protect your extension, even if you avoid binding redirects forcing your extension to use the assembly that NuGet ships, rather than what your assembly compiled against (although that's going to fail for a different reason).

File: src/NuGet.Clients/NuGet.VisualStudio.Interop/NuGet.VisualStudio.Interop.csproj (hunk 2 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.VisualStudio.Interop/NuGet.VisualStudio.Interop.csproj:7]: Oh ok, I thought it stopped packing n.vs.interop. I'm OK with this change

**Review by dominoFire — APPROVED** [review]
Signing off. Please, make sure if other VS-NuGet related packages (N.PM.UI, N.PM, ...) needs to change to this version

**Review by nkolev92 — APPROVED** [review]

**Review by kartheekp-ms — APPROVED** [review]
Thank you, Andy, for adding great details to this PR in description and other comments.

I think after this change
- `NuGet version` in VS will change in PMC. For e.g. it will display `NuGet Version 17.1.0`
- I see that `NuGetSDKVersion` property is added to `buildinfo.json`. After we change VS insertion script the next attempt to insert into `dev17` may fail because the new property will not be added to the json file. Maybe we need to cherry-pick this commit to `release-6.0.x` branch.

I think the root cause appears to be with the way how we do VS insertion. When we do a VS insertion, we don't publish `NuGet.VisualStudio` and `NuGet.VisualStudio.Contracts` nupkgs to `nuget.org` and we publish these packages along with other `Core` packages which has high build number.

**Comment by nkolev92** [comment]
> NuGet version in VS will change in PMC. For e.g. it will display NuGet Version 17.1.0

Pretty sure this won't be the case, but @zivkan can maybe clarify it.

**Comment by zivkan** [comment]
> NuGet version in VS will change in PMC. For e.g. it will display NuGet Version 17.1.0

No, it won't.

That message is defined here: https://github.com/NuGet/NuGet.Client/blob/22c8adc5dc236b32e5fafb6899f0acd193aaee92/src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs#L633

and `_nugetHost.Version` is defined here: https://github.com/NuGet/NuGet.Client/blob/22c8adc5dc236b32e5fafb6899f0acd193aaee92/src/NuGet.Clients/NuGetConsole.Host.PowerShell/NuGetPSHost.cs#L77

Therefore the version number displayed is the assembly version of NuGet.Console.Host.PowerShell.dll, which is not modified by this PR.

> Maybe we need to cherry-pick this commit to release-6.0.x branch.

I've already updated the release pipeline to fall back to current behaviour when the property doesn't exist:

```powershell
$buildInfo = Get-Content buildInfo.json | ConvertFrom-Json
if (Get-Member -InputObject $buildInfo -Name NuGetSdkVsVersion) {
  $assemblyVersion = $buildInfo.NuGetSdkVsVersion + ".0"
} else {
  $assemblyVersion = $buildInfo.BuildNumber
}
```

Therefore, there's no need to cherry-pick to old branches, or change the package version of NuGet.VisualStudio or NuGet.VisualStudio.Contracts if we end up servicing old releases.
